### PR TITLE
aws-asg: support ec2role credentials if AWS access keypair is empty or omitted

### PR DESCRIPTION
**Pull request by adrianlop**
# Description
This PR adds support for the `aws-asg` target to use EC2 role credentials if the AWS access keypair config params are empty `""` or are omitted in the autoscaler configfile (fixes https://github.com/hashicorp/nomad-autoscaler/issues/557)
It also upgrades the AWS Go SDK v2 from 0.23 to 1.13 (which had breaking changes, that are resolved in the PR too)

I've tested this code in the following scenarios:
- running the autoscaler agent in both EC2 instance as a binary, and running as a Docker container inside an EC2 instance with the appropriate IAM role associated to the instance ✅ 
- omitting the `aws_access_key_id` and `aws_secret_access_key` config params, and also declaring them with empty strings `""` ✅ 
- previous behaviour (access keypair config params having actual values) keeps working

### CODE REVIEW

**Review by lgfa29 — APPROVED** [review]
LGTM! Thanks for handling the SKD upgrade as well 🎉 

I added some minor comments, but no blockers. I also ran our [AWS demo](https://learn.hashicorp.com/tutorials/nomad/horizontal-cluster-scaling?in=nomad/autoscaler) and everything worked.

File: plugins/builtin/target/aws-asg/plugin/aws.go [hunk 1]
```diff
@@ -48,24 +51,28 @@ func (t *TargetPlugin) setupAWSClients(config map[string]string) error {
 	// supplied configuration. In order to use these static credentials both
 	// the access key and secret key need to be present; the session token is
 	// optional.
+	// If not found, EC2RoleProvider will be instantiated instead.
 	keyID, idOK := config[configKeyAccessID]
 	secretKey, keyOK := config[configKeySecretKey]
 	session := config[configKeySessionToken]
 
-	if idOK && keyOK {
+	if idOK && keyOK && keyID != "" && secretKey != "" {
```
[the remark this answers is not in the document]
Comment (adrianlop) [reply plugins/builtin/target/aws-asg/plugin/aws.go:59]: ah true, thanks! will commit that 👍

File: plugins/builtin/target/aws-asg/plugin/aws.go [hunk 2]
```diff
@@ -48,24 +51,28 @@ func (t *TargetPlugin) setupAWSClients(config map[string]string) error {
...
 		t.logger.Trace("setting AWS access credentials from config map")
-		cfg.Credentials = aws.NewStaticCredentialsProvider(keyID, secretKey, session)
+		cfg.Credentials = credentials.NewStaticCredentialsProvider(keyID, secretKey, session)
+	} else {
+		t.logger.Trace("AWS access credentials empty - using EC2 instance role credentials instead")
+		cfg.Credentials = aws.NewCredentialsCache(ec2rolecreds.New())
```
[the remark this answers is not in the document]
Comment (adrianlop) [reply plugins/builtin/target/aws-asg/plugin/aws.go:64]: when I was testing, the auth errors were clear enough to me
but maybe we could try something like this (after the if/else stmt), if we want to err earlier:
```
_, err = cfg.Credentials.Retrieve(context.TODO())
if err != nil {
	return fmt.Errorf("failed to authenticate with AWS: %v", err)
}
```

[the remark this answers is not in the document]
Comment (lgfa29) [reply plugins/builtin/target/aws-asg/plugin/aws.go:64]: Ah OK. If the logged message was good then I think we don't have to worry about it 👍

**Comment by peter-lockhart-pub** [comment]
Hi, I stumbled across this change as we were looking to upgrade the version of the Autoscaler we use. v0.3.6 (which includes this change) no longer works for us, while v0.3.5 does work. Perhaps there is some documentation that needs updated along with this change?

We run the autoscaler on k8s, and use `eks.amazonaws.com/role-arn: "arn:aws:iam::<account>:role/<autoscaler role>"` , where that role has the permissions laid out in https://www.nomadproject.io/tools/autoscaling/plugins/target/aws-asg.

On the pod, this means in practice it's ENV variables are like 
```
AWS_DEFAULT_REGION: <region>
AWS_REGION: <region>
AWS_ROLE_ARN: arn:aws:iam::<account>:role/<autoscaler role>
AWS_WEB_IDENTITY_TOKEN_FILE: /var/run/secrets/eks.amazonaws.com/serviceaccount/token
```

With v0.3.6, we can see from the logs that it now using the incorrect role:
`2022-03-03T16:10:06.782Z [WARN]  policy_manager.policy_handler: failed to get target status: policy_id=73402811-562f-43ab-b7fe-9fe0d62e1536 error="failed to describe AWS Autoscaling Group: operation error Auto Scaling: DescribeAutoScalingGroups, https response error StatusCode: 403, RequestID: 0fbe3a1c-9983-4e56-9f82-503a836ab1fb, api error AccessDenied: User: arn:aws:sts::<account id>:assumed-role/<not the right role>/i-<redacted> is not authorized to perform: autoscaling:DescribeAutoScalingGroups because no identity-based policy allows the autoscaling:DescribeAutoScalingGroups action"`

This does not occur when we run v0.3.5.

**Comment by lgfa29** [comment]
@peter-lockhart-pub would you mind opening a separate issue to help us track this problem?

Thanks!

**Pull request merged**